torchx/c_src/torchx.cpp

```diff
@@ -0,0 +1,531 @@
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::scalar_tensor(scalar, type));
+}
+
+NIF(randint)
+{
+  PARAM(0, int64_t, min);
+  PARAM(1, int64_t, max);
+  SHAPE_PARAM(2, shape);
+  TYPE_PARAM(3, type);
+
+  TENSOR(at::randint(min, max, shape, type));
+}
+
+NIF(rand)
+{
+  PARAM(0, double, min);
+  PARAM(1, double, max);
+  SHAPE_PARAM(2, shape);
+  TYPE_PARAM(3, type);
+
+  TENSOR(min + at::rand(shape, type) * (max - min));
+}
+
+NIF(normal)
+{
+  PARAM(0, double, mean);
+  PARAM(1, double, std);
+  SHAPE_PARAM(2, shape);
+
+  TENSOR(at::normal(mean, std, shape));
+}
+
+NIF(arange)
+{
+  PARAM(0, int64_t, start);
+  PARAM(1, int64_t, end);
+  PARAM(2, int64_t, step);
+  TYPE_PARAM(3, type);
+
+  if (argc == 5)
+  {
+    SHAPE_PARAM(4, shape);
+    TENSOR(at::reshape(at::arange((double)start, (double)end, (double)step, type), shape));
+  }
+  else
+  {
+    TENSOR(at::arange((double)start, (double)end, (double)step, type));
+  }
+}
+
+NIF(reshape)
+{
+  TENSOR_PARAM(0, t);
+  SHAPE_PARAM(1, shape);
+
+  TENSOR(at::reshape(*t, shape));
+}
+
+NIF(to_type)
+{
+  TENSOR_PARAM(0, t);
+  TYPE_PARAM(1, type);
+
+  TENSOR(t->toType(type));
+}
+
+NIF(squeeze)
+{
+  TENSOR_PARAM(0, t);
+
+  if (argc == 2)
+  {
+    PARAM(1, int64_t, dim);
+    TENSOR(at::squeeze(*t, dim));
+  }
+  else
+    TENSOR(at::squeeze(*t));
+}
+
+NIF(broadcast_to)
+{
+  TENSOR_PARAM(0, t);
+  SHAPE_PARAM(1, shape);
+
+  TENSOR(at::broadcast_to(*t, shape).clone());
+}
+
+NIF(transpose)
+{
+  TENSOR_PARAM(0, t);
+  PARAM(1, int64_t, dim0);
+  PARAM(2, int64_t, dim1);
+
+  TENSOR(at::transpose(*t, dim0, dim1));
+}
+
+NIF(ones)
+{
+  SHAPE_PARAM(0, shape);
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::ones(shape, type));
+}
+
+NIF(eye)
+{
+  PARAM(0, int64_t, size);
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::eye(size, type));
+}
+
+#define BINARY_OP(OP)   BINARY_OP2(OP, OP)
+
+#define BINARY_OP2(OP, NATIVE_OP)   \
+  NIF(OP)               \
+  {                     \
+    TENSOR_PARAM(0, a); \
+    TENSOR_PARAM(1, b); \
+                        \
+    TENSOR(at::NATIVE_OP(*a, *b)); \
+  }
+
+#define BINARY_OPB(OP)   \
+  NIF(OP)               \
+  {                     \
+    TENSOR_PARAM(0, a); \
+    TENSOR_PARAM(1, b); \
+                        \
+    nx::nif::ok(env, nx::nif::make(env, at::OP(*a, *b))); \
+  }
+
+
+#define UNARY_OP(OP) UNARY_OP2(OP, OP)
+
+#define UNARY_OP2(OP, NATIVE) \
+  NIF(OP)               \
+  {                     \
+    TENSOR_PARAM(0, a); \
+    TENSOR(at::NATIVE(*a)); \
+  }
+
+
+
+BINARY_OP(bitwise_and)
+BINARY_OP(bitwise_or)
+BINARY_OP(bitwise_xor)
+BINARY_OP2(left_shift, __lshift__)
+BINARY_OP2(right_shift, __rshift__)
+
+BINARY_OP2(equal, eq)
+BINARY_OP(not_equal)
+BINARY_OP(greater)
+BINARY_OP(less)
+BINARY_OP(greater_equal)
+BINARY_OP(less_equal)
+
+BINARY_OP(logical_and)
+BINARY_OP(logical_or)
+BINARY_OP(logical_xor)
+
+BINARY_OP(add)
+BINARY_OP(subtract)
+BINARY_OP(divide)
+BINARY_OP(remainder)
+BINARY_OP(multiply)
+BINARY_OP2(power, float_power)
```

Comment:
Perhaps, to avoid future confusion, at the NIF level we should name the functions after libtorch API, and not after the Nx one?

Reply:
On the one hand — yes. And it would simplify C++ code a bit.
On the other — we would not be able to do binary and unary operations code generation in cycles in backend, if names differ. But we can replace function atoms with tuples in operations lists. I'll have a look at that.

For other functions (like `arange`, `randint` etc.) I use original pytorch names.

Reply:
>  On the other — we would not be able to do binary and unary operations code generation in cycles in backend, if names differ. But we can replace function atoms with tuples in operations lists. I'll have a look at that.

Ah, I see. Ok, let's go with the consistent Nx naming then!